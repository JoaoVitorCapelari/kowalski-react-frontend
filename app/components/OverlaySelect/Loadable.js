/**
 *
 * Asynchronously loads the component for OverlaySelect
 *
 */

import Loadable from 'react-loadable';

export default Loadable({
  loader: () => import('./index'),
  loading: () => null,
});
