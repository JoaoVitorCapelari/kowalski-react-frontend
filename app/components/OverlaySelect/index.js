/**
*
* OverlaySelect
*
*/

import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';

const Overlay = styled.div`
  background-color: #fff;
  width: 100%;
  height: 100%;
  z-index: 10;
  top: 0;
  left: 0;
  position: fixed;
  padding: 20px;
  padding-left: 30px;
  padding-right: 30px;
  opacity: 0.95;
  display: flex;
  flex-direction: column;
  align-items: center;
`;

const OverlaySelectInnerContentContainer = styled.div`
  margin-top: 1rem;
`;

class OverlaySelect extends React.Component { // eslint-disable-line react/prefer-stateless-function
  componentDidMount() {
    document.addEventListener('keydown', this.dismissOnEsc, false);
  }

  componentWillUnmount() {
    document.removeEventListener('keydown', this.dismissOnEsc, false);
  }

  dismissOnEsc = (event) => {
    if (event.keyCode === 27) {
      this.props.onDismiss();
    }
  };

  render() {
    const { title, children, opened, onOptionSelect, onDismiss, ...other } = this.props;

    if (!opened) {
      return <div></div>;
    }

    const wrappedOnOptionSelect = (value) => {
      onOptionSelect(value);
      this.props.onDismiss();
    };

    return (
      <Overlay>
        <h3>{title}</h3>
        <OverlaySelectInnerContentContainer>
          {React.Children.map(children, ((child) =>
            React.cloneElement(child, {
              onOptionSelect: wrappedOnOptionSelect,
              ...other,
            })
          ))}
        </OverlaySelectInnerContentContainer>
      </Overlay>
    );
  }
}

OverlaySelect.propTypes = {
  children: PropTypes.any,
  onOptionSelect: PropTypes.func.isRequired,
  selectedOption: PropTypes.any,
  title: PropTypes.string,
  opened: PropTypes.bool.isRequired,
  onDismiss: PropTypes.func.isRequired,
};

export default OverlaySelect;
