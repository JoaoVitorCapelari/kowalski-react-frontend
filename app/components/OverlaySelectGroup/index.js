/**
*
* OverlaySelectGroup
*
*/

import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';

const OverlaySelectGroupContainer = styled.div`
  margin-top: 15px;
`;

function OverlaySelectGroup({ children, ...other }) {
  return (
    <OverlaySelectGroupContainer>
      {React.Children.map(children, ((child) => React.cloneElement(child, other)))}
    </OverlaySelectGroupContainer>
  );
}

OverlaySelectGroup.propTypes = {
  children: PropTypes.any,
  onOptionSelect: PropTypes.func,
  selectedOption: PropTypes.any,
};

export default OverlaySelectGroup;
