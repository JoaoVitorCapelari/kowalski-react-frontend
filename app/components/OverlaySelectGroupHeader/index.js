/**
*
* OverlaySelectGroupHeader
*
*/

import React from 'react';
import styled from 'styled-components';
import PropTypes from 'prop-types';

const Title = styled.h4`
`;

function OverlaySelectGroupHeader(props) {
  return (
    <Title>
      { props.children }
    </Title>
  );
}

OverlaySelectGroupHeader.propTypes = {
  children: PropTypes.any,
};

export default OverlaySelectGroupHeader;
