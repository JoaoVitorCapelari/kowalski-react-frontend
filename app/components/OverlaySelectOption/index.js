/**
*
* OverlaySelectOption
*
*/

import React from 'react';
import styled from 'styled-components';
import PropTypes from 'prop-types';

const Option = styled.a`
  display: block;
`;

const SelectedOption = styled.a`
  display: block;
  color: blue;
`;

function OverlaySelectOption({ onOptionSelect, selectedOption, children, value }) {
  const OptionElement = selectedOption === value || children
    ? SelectedOption
    : Option;
  const finalValue = value || children;

  return (
    <OptionElement onClick={() => onOptionSelect(finalValue)}>
      { children }
    </OptionElement>
  );
}

OverlaySelectOption.propTypes = {
  onOptionSelect: PropTypes.func,
  selectedOption: PropTypes.any,
  children: PropTypes.any,
  value: PropTypes.any,
};

export default OverlaySelectOption;
