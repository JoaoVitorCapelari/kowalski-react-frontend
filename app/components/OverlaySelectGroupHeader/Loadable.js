/**
 *
 * Asynchronously loads the component for OverlaySelectGroupHeader
 *
 */

import Loadable from 'react-loadable';

export default Loadable({
  loader: () => import('./index'),
  loading: () => null,
});
