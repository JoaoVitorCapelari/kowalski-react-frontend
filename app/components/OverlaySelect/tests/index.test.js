import React from 'react';
import { mount } from 'enzyme';

import OverlaySelect from '../index';
import OverlaySelectGroup from '../../OverlaySelectGroup';
import OverlaySelectGroupHeader from '../../OverlaySelectGroupHeader';
import OverlaySelectOption from '../../OverlaySelectOption';

describe('<OverlaySelect />', () => {
  let children;
  let onOptionSelect;
  let selectedOption;
  let title;
  let opened;
  let onDismiss;
  let renderedComponent;

  beforeEach(() => {
    onOptionSelect = jest.fn();
    onDismiss = jest.fn();
    title = 'My Select is here!';
  });

  const renderComponent = () => mount(
    <OverlaySelect {...{ onOptionSelect, selectedOption, title, opened, onDismiss }}>
      { children }
    </OverlaySelect>
  );

  describe('when opened', () => {
    beforeEach(() => {
      opened = true;
    });

    describe('when having some children', () => {
      beforeEach(() => {
        children = [
          <OverlaySelectGroup key={1}>
            <OverlaySelectGroupHeader title="teste" />
            <OverlaySelectOption value={1} />
          </OverlaySelectGroup>,
          <OverlaySelectGroup key={2}>
            <OverlaySelectGroupHeader title="teste2" />
            <OverlaySelectOption>2</OverlaySelectOption>
          </OverlaySelectGroup>,
        ];
        selectedOption = 2;
        renderedComponent = renderComponent();
      });

      it('renders', () => {
        expect(renderedComponent.html()).toMatchSnapshot();
        expect(renderedComponent.html()).toContain(title);
      });

      it('unmounts', () => {
        renderedComponent.unmount();
      });

      it('calls onDismiss on esc press', () => {
        renderedComponent.instance().dismissOnEsc({ keyCode: 10 });
        expect(onDismiss).not.toHaveBeenCalled();
        renderedComponent.instance().dismissOnEsc({ keyCode: 27 });
        expect(onDismiss).toHaveBeenCalled();
      });

      it('callbacks onOptionSelect when event is fired from children', () => {
        renderedComponent.find('a').first().simulate('click');
        expect(onOptionSelect).toHaveBeenCalledWith(1);
      });
    });

    describe('without children', () => {
      beforeEach(() => {
        children = [];
        renderedComponent = renderComponent();
      });

      it('renders', () => {
        expect(renderedComponent.html()).toMatchSnapshot();
        expect(renderedComponent.html()).toContain(title);
      });
    });
  });

  describe('when not opened', () => {
    beforeEach(() => {
      opened = false;
      renderedComponent = renderComponent();
    });

    it('does not render', () => {
      expect(renderedComponent.html()).toMatchSnapshot();
      expect(renderedComponent.html()).not.toContain(title);
    });
  });
});
