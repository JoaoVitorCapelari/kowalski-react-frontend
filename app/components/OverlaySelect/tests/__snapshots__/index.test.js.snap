// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<OverlaySelect /> Expect to have unit tests specified 1`] = `"<div class=\\"OverlaySelect__Overlay-kpvDhg gGlSub\\"><h3></h3><div class=\\"OverlaySelect__OverlaySelectInnerContentContainer-dZagTg gNCBAw\\"></div></div>"`;
